pkg/controller/garbagecollector/garbagecollector.go

```diff
@@ -171,6 +205,109 @@ func referencesDiffs(old []metatypes.OwnerReference, new []metatypes.OwnerRefere
 	return added, removed
 }
 
+func shouldFinalize(e event, accessor meta.Object) bool {
+	// The delta_fifo may combine the creation and update of the object into one
+	// event, so we need to check AddEvent as well.
+	if e.oldObj == nil {
+		if accessor.GetDeletionTimestamp() == nil {
+			return false
+		}
+	} else {
+		oldAccessor, err := meta.Accessor(e.oldObj)
+		if err != nil {
+			utilruntime.HandleError(fmt.Errorf("cannot access oldObj: %v", err))
+			return false
+		}
+		// ignore the event if it's not updating DeletionTimestamp from non-nil to nil.
+		if accessor.GetDeletionTimestamp() == nil || oldAccessor.GetDeletionTimestamp() != nil {
+			return false
+		}
+	}
+	return sets.NewString(accessor.GetFinalizers()...).Has(OrphanFinalizerID)
+}
+
+func (gc *GarbageCollector) orphan(owner *node) {
+	// we don't need to lock each element, because they never get updated
+	owner.dependentsLock.RLock()
+	dependents := make([]*node, 0, len(owner.dependents))
+	for dependent := range owner.dependents {
+		dependents = append(dependents, dependent)
+	}
+	owner.dependentsLock.RUnlock()
+
+	// patch operation has less conflicts, so make its max retries lower
+	patchMaxRetries := 3
+	var failedDependents []objectReference
+	for _, dependent := range dependents {
+		deleteOwnerRefPatch := fmt.Sprintf(`{"metadata":{"ownerReferences":[{"$patch":"delete","uid":"%s"}]}}`, owner.identity.UID)
+		patched := false
+		for count := 0; count < patchMaxRetries; count++ {
+			_, err := gc.patchObject(dependent.identity, []byte(deleteOwnerRefPatch))
+			if err == nil {
+				patched = true
+				break
+			} else {
+				utilruntime.HandleError(fmt.Errorf("orphaning %s failed with %v", dependent.identity, err))
+			}
+		}
+		if !patched {
+			failedDependents = append(failedDependents, dependent.identity)
+		}
+	}
+	if len(failedDependents) != 0 {
+		utilruntime.HandleError(fmt.Errorf("failed to orphan dependents %v, please manually orphan them and finalize the owner %s", failedDependents, owner.identity))
```

Comment:
I suggest you think of orphaning as a status for the node. Like it gets put into that state and we keep trying to remove the references from the dependents until we succeed. Not necessarily in a single loop like you have here, but like every time the object gets looked at in the queue it should drop into this function. And if it fails here it should get re-enqueued so that it gets another time slice in a bit.